src/client/opamConfigCommand.ml

```diff
@@ -382,6 +382,87 @@ let variable gt v =
+  let open OpamStd.Option.Op in
+  match OpamStd.List.find_opt ((=) field @* fst) OpamFile.Config.fields with
+  | None ->
+    OpamConsole.error_and_exit `Bad_arguments
+      "Field %s doesn't exist" (OpamConsole.colorise `underline field)
+  | Some (_, pp) ->
+    let allowed =
+      let in_config = OpamInitDefaults.init_config () in
+      [ "download-command",
+        OpamFile.Config.with_dl_tool_opt
+          (OpamFile.InitConfig.dl_tool in_config
+           ++ OpamFile.Config.dl_tool OpamFile.Config.empty);
+        "download-jobs",
+        OpamFile.Config.with_dl_jobs
+          (OpamFile.InitConfig.dl_jobs in_config
+           +! OpamFile.Config.dl_jobs OpamFile.Config.empty);
+        "jobs", OpamFile.Config.with_jobs
+          (OpamFile.InitConfig.jobs in_config
+           +! OpamFile.Config.jobs OpamFile.Config.empty);
+        "best-effort-prefix-criteria",
+        OpamFile.Config.with_best_effort_prefix_opt
+          (OpamFile.Config.best_effort_prefix OpamFile.Config.empty);
+        "solver",
+        OpamFile.Config.with_solver_opt
+          (OpamFile.InitConfig.solver in_config
+           ++ OpamFile.Config.solver OpamFile.Config.empty);
+        "global-variables",
+        OpamFile.Config.with_global_variables
+          (OpamFile.InitConfig.global_variables in_config);
+        "eval-variables",
+        OpamFile.Config.with_eval_variables
+          (OpamFile.InitConfig.eval_variables in_config);
+        "repository-validation-command",
+        OpamFile.Config.with_validation_hook_opt
+          (OpamFile.Config.validation_hook OpamFile.Config.empty);
+      ] @ List.map (fun f ->
+          f, OpamFile.Config.with_criteria
+            (OpamFile.Config.criteria OpamFile.Config.empty))
+        [ "solver-criteria";
+          "solver-upgrade-criteria";
+          "solver-fixup-criteria" ]
+    in
+    match OpamStd.List.find_opt ((=) field @* fst) allowed with
+    | None ->
+      OpamConsole.error_and_exit `Bad_arguments
+        "Field %s is not modifiable" (OpamConsole.colorise `underline field)
+    | Some (_, set_default) ->
+      let config = gt.config in
+      let new_config =
+        match value with
+        | Some value ->
+          (try
+             let value = OpamParser.value_from_string value "<command-line>" in
+             OpamPp.parse ~pos:OpamTypesBase.pos_null pp (config, Some value)
+           with
+           | OpamPp.Bad_format (_,str) ->
+             OpamConsole.error_and_exit `Bad_arguments
+               "Parse error on %s value '%s': %s"
+               (OpamConsole.colorise `underline field)
+               value str)
+        | None ->
+          if OpamConsole.confirm ~default:false
+              "Are you sure you want to revert %s value"
+              (OpamConsole.colorise `underline field)
+          then
+            set_default config
+          else config
+      in
+      let _, _gt =
+        OpamGlobalState.with_write_lock gt @@ fun gt ->
```

Comment:
Don't use lock upgrade.
Also, it's useless to lock just for the write (the files are individually locked already at the OpamFile layer when writing), what we need it to keep the lock for the whole time between load and write, so that we are sure not to interfere with concurrent writes.

Reply:
After some discussion, `OpamConfigCommand` function will ben updated to always take & returns the state